src/session-pool.js

```diff
@@ -435,252 +366,370 @@ SessionPool.prototype.requestStream = function(config) {
+
+  this.pendingCreates_ += 1;
+
+  return this.requestQueue_
+    .add(function() {
+      return session.create();
+    })
+    .then(
+      function() {
+        self.available_.push(session);
+        self.pendingCreates_ -= 1;
+        return session;
+      },
+      function(err) {
+        self.pendingCreates_ -= 1;
+        throw err;
+      }
+    );
+};
+
+/**
+ * Attempts to create a session but emits any errors that occur.
+ *
+ * @private
+ *
+ * @emits SessionPool#available
+ * @emits SessionPool#error
+ * @param {string} type The desired type to create.
+ * @returns {Promise}
+ */
+SessionPool.prototype.createSessionInBackground_ = function(type) {
+  var self = this;
 
-  return session.create().then(
+  return this.createSession_(type).then(
     function() {
-      return session;
+      self.emit('available');
     },
     function(err) {
-      return err;
+      self.emit('error', err);
     }
   );
 };
 
 /**
- * Creates a Transaction, stubs a destroy method to return the transaction
- * back into the pool.
+ * Attempts to delete a session, optionally creating a new one of the same type
+ * if the pool is still open and we're under the configured min value.
  *
  * @private
  *
- * @return {Transaction}
+ * @fires SessionPool#destroy
+ * @fires SessoinPool#error
+ * @param {Session} session The session to delete.
+ * @returns {Promise}
  */
-SessionPool.prototype.createTransaction_ = function(session, options) {
+SessionPool.prototype.destroySession_ = function(session) {
   var self = this;
+  var index = this.available_.indexOf(session);
 
-  var transaction = session.transaction(options);
-  var end = transaction.end.bind(transaction);
+  this.available_.splice(index, 1);
+  this.emit('destroy');
 
-  transaction.end = function(callback) {
-    self.release(session);
-    end(callback);
-  };
+  if (this.isOpen && this.size() < this.options.min) {
+    this.createSessionInBackground_(session.type);
+  }
 
-  return transaction;
+  return this.requestQueue_
+    .add(function() {
+      return session.delete();
+    })
+    .catch(function(err) {
+      self.emit('error', err);
+    });
 };
 
 /**
- * Create a session, then begin a new transaction.
+ * Deletes idle sessions that exceed the maxIdle configuration.
  *
  * @private
  *
- * @returns {Promise} - Resolves to {Session}.
+ * @returns {Promise}
  */
-SessionPool.prototype.createWriteSession_ = function() {
+SessionPool.prototype.evictIdleSessions_ = function() {
   var self = this;
+  var evicted = [];
 
-  return this.createSession_().then(function(session) {
-    if (session instanceof Error) {
-      return Promise.resolve(session);
-    }
+  var maxIdle = this.options.maxIdle;
+  var min = this.options.min;
 
-    var transaction = self.createTransaction_(session);
+  var idle = this.getIdleSessions_();
+  var count = idle.length;
 
-    session.isWriteSession_ = true;
-    session.transaction_ = transaction;
+  while (count-- > maxIdle && count >= min) {
```

Comment:
I think you should compare size() to min?

Reply:
Done! PTAL